packages/firestore/test/unit/util/bundle.test.ts

```diff
@@ -0,0 +1,354 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { expect } from 'chai';
+import { Bundle, toReadableStream } from '../../../src/util/bundle';
+import { isNode } from '../../util/test_platform';
+
+function readableStreamFromString(
+  content: string,
+  bytesPerRead: number
+): ReadableStream {
+  return toReadableStream(new TextEncoder().encode(content), bytesPerRead);
+}
+
+function lengthPrefixedString(o: {}): string {
+  const str = JSON.stringify(o);
+  const l = new TextEncoder().encode(str).byteLength;
+  return `${l}${str}`;
+}
+
+// eslint-disable-next-line no-restricted-properties
+(isNode() ? describe.skip : describe)('readableStreamFromString()', () => {
+  it('returns stepping readable stream', async () => {
+    const encoder = new TextEncoder();
+    const s = readableStreamFromString('0123456789', 4);
+    const r = s.getReader();
+
+    let result = await r.read();
+    expect(result.value).to.deep.equal(encoder.encode('0123'));
+    expect(result.done).to.be.false;
+
+    result = await r.read();
+    expect(result.value).to.deep.equal(encoder.encode('4567'));
+    expect(result.done).to.be.false;
+
+    result = await r.read();
+    expect(result.value).to.deep.equal(encoder.encode('89'));
+    expect(result.done).to.be.false;
+
+    result = await r.read();
+    expect(result.value).to.be.undefined;
+    expect(result.done).to.be.true;
+  });
+});
+
+// eslint-disable-next-line no-restricted-properties
+(isNode() ? describe.skip : describe)('Bundle ', () => {
+  genericBundleReadingTests(1);
+  genericBundleReadingTests(4);
+  genericBundleReadingTests(64);
+  genericBundleReadingTests(1024);
+});
+
+function genericBundleReadingTests(bytesPerRead: number): void {
+  // Setting up test data.
+  const meta = {
```

Comment:
Would it make sense to check this as a test resource and to read it from a file instead of setting this up here in this test? That makes the test much shorter (and more readable) and matches the expected usage pattern.

Reply:
It definitely makes it shorter, but might not be as more readable as the size reduction? Because readers now have to read the built bundle content to make sure the tests are doing the right thing?

Reply:
I see your point. However, I think this test is less readable precisely because of this setup. I can go either way here and defer to your judgement.